File: engines/user-feature-extraction-engine-new/src/main/scala/io/sugo/pio/engine/userFeatureExtractionNew/engine/UserFeatureExtractionEngine.scala
package io.sugo.pio.engine.userFeatureExtractionNew.engine

import com.typesafe.config.Config
import io.sugo.pio.engine.training._
import io.sugo.pio.engine.userFeatureExtractionNew.data.{UserFeatureExtractionModelData, UserFeatureExtractionPrepareData, UserFeatureExtractionTrainingData}
import io.sugo.pio.engine.userFeatureExtractionNew.eval.{UserFeatureExtractionEvalActualResult, UserFeatureExtractionEvalIndicators, UserFeatureExtractionEvalQuery}
import io.sugo.pio.engine.userFeatureExtractionNew.module.{UserFeatureExtractionAlgorithm, UserFeatureExtractionDatasource, UserFeatureExtractionModel, UserFeatureExtractionPreparator}

/**
  * Created by penghuan on 2017/4/24.
  */
class UserFeatureExtractionEngine(config: Config, engineParams: EngineParams)
  extends Engine [
    UserFeatureExtractionTrainingData,
    UserFeatureExtractionPrepareData,
    UserFeatureExtractionModelData,
    UserFeatureExtractionEvalQuery,
    UserFeatureExtractionEvalActualResult,
    UserFeatureExtractionEvalIndicators
  ] (engineParams) {

  override protected def createDatasource(params: Params): DataSource[UserFeatureExtractionTrainingData, UserFeatureExtractionEvalQuery, UserFeatureExtractionEvalActualResult] = {
    new UserFeatureExtractionDatasource(config)
  }

  override protected def createPreparator(params: Params): Preparator[UserFeatureExtractionTrainingData, UserFeatureExtractionPrepareData] = {
    new UserFeatureExtractionPreparator
  }

  override protected def createAlgorithm(params: Params): Algorithm[UserFeatureExtractionPrepareData, UserFeatureExtractionModelData] = {
    new UserFeatureExtractionAlgorithm
  }

  override protected def createModel(): Model[UserFeatureExtractionModelData] = {
    new UserFeatureExtractionModel(config)
  }

  override protected def createEval(): Evalution[UserFeatureExtractionModelData, UserFeatureExtractionEvalQuery, UserFeatureExtractionEvalActualResult, UserFeatureExtractionEvalIndicators] = {
    null
  }
}

File: engines/user-feature-extraction-engine/src/main/java/io/sugo/pio/engine/userFeatureExtraction/UserFeatureExtractionEngine.java
package io.sugo.pio.engine.userFeatureExtraction;

import io.sugo.pio.engine.data.input.BatchEventHose;
import io.sugo.pio.engine.data.input.PropertyHose;
import io.sugo.pio.engine.data.output.Repository;
import io.sugo.pio.engine.training.*;
import io.sugo.pio.engine.userFeatureExtraction.data.UserFeatureExtractionModelData;
import io.sugo.pio.engine.userFeatureExtraction.data.UserFeatureExtractionPrepareData;
import io.sugo.pio.engine.userFeatureExtraction.data.UserFeatureExtractionTrainingData;
import io.sugo.pio.engine.userFeatureExtraction.engine.UserFeatureExtractionAlgorithm;
import io.sugo.pio.engine.userFeatureExtraction.engine.UserFeatureExtractionDatasource;
import io.sugo.pio.engine.userFeatureExtraction.engine.UserFeatureExtractionModel;
import io.sugo.pio.engine.userFeatureExtraction.engine.UserFeatureExtractionPreparator;
import io.sugo.pio.engine.userFeatureExtraction.eval.UserFeatureExtractionEvalActualResult;
import io.sugo.pio.engine.userFeatureExtraction.eval.UserFeatureExtractionEvalIndicators;
import io.sugo.pio.engine.userFeatureExtraction.eval.UserFeatureExtractionEvalQuery;

/**
 * Created by penghuan on 2017/4/7.
 */
public class UserFeatureExtractionEngine extends Engine<UserFeatureExtractionTrainingData, UserFeatureExtractionPrepareData, UserFeatureExtractionModelData, UserFeatureExtractionEvalQuery, UserFeatureExtractionEvalActualResult, UserFeatureExtractionEvalIndicators> {

    private final BatchEventHose batchEventHose;
    private final PropertyHose propertyHose;
    private final Repository repository;
    private final EngineParams engineParams;

    public UserFeatureExtractionEngine(
            PropertyHose propertyHose,
            BatchEventHose batchEventHose,
            Repository repository,
            EngineParams engineParams) {
        super(engineParams);
        this.batchEventHose = batchEventHose;
        this.propertyHose = propertyHose;
        this.repository = repository;
        this.engineParams = engineParams;
    }

    @Override
    protected DataSource<UserFeatureExtractionTrainingData, UserFeatureExtractionEvalQuery, UserFeatureExtractionEvalActualResult> createDatasource(Params params) {
        return new UserFeatureExtractionDatasource(propertyHose, batchEventHose);
    }

    @Override
    protected Preparator<UserFeatureExtractionTrainingData, UserFeatureExtractionPrepareData> createPreparator(Params params) {
        return new UserFeatureExtractionPreparator();
    }

    @Override
    protected Algorithm<UserFeatureExtractionPrepareData, UserFeatureExtractionModelData> createAlgorithm(Params params) {
        return new UserFeatureExtractionAlgorithm();
    }

    @Override
    protected Model<UserFeatureExtractionModelData> createModel() {
        return new UserFeatureExtractionModel(repository);
    }

    @Override
    protected Evalution<UserFeatureExtractionModelData, UserFeatureExtractionEvalQuery, UserFeatureExtractionEvalActualResult, UserFeatureExtractionEvalIndicators> createEval() {
        return null;
    }
}
